pkg/kubelet/cm/cpumanager/cpu_manager.go

```diff
@@ -343,11 +359,14 @@ func (m *manager) removeStaleState() {
 	defer m.Unlock()
 
 	// Get the list of active pods.
-	activePods := m.activePods()
+	activeAndAdmittedPods := m.activePods()
+	for _, pod := range m.admittedPods {
+		activeAndAdmittedPods = append(activeAndAdmittedPods, pod)
+	}
```

Comment:
I think this is only partially correct.  You need to prune the admitted pods map to only include those that are still in either config or pod worker but not in both (a pod that is in the pod worker is by definition admitted).  You also want to exclude pods that are no longer running due to termination (which means we will never start the container again, thus we can reuse those CPUs).  You then need to take that pruned list of admitted pods and add it to active pods to get the list of "pods that should be considered to be allocated guaranteed resources"

I think we'll need to add a method on kubelet to make this easier - looking at what that would be right now.

Reply:
Looking at this, I think we have a bug in that these components need to know about pods that are terminating that have been force deleted (which activePods() does not include).  @Random-Liu I think that components like the cpu_manager, memory_manager, qos_container_manager, and device_manager actually need to know pods that have been force deleted but are still running.  To do that, I'm going to need to enable the pod worker to share that info safely (and consistently, i think).

Reply:
I'm not sure what you suggest here is necessary. As you mentioned in your comment [here](https://github.com/kubernetes/kubernetes/issues/103952#issuecomment-888387811), all we should really need to care about is whether the pod is:

(1) in the active pods list; or
(2) some pod currently waiting to be admitted

Assuming logic exists to _atomically_ add an admitted pod to the active pod list before starting the next admission loop, I believe this should be sufficient.

We just then need to track a single variable to hold the "currently being admitting" pod that gets overwritten each time a new iteration of the admit loop starts.

The check in `removeStaleState()` would then remove any state not currently associated with a pod in the active list + this new variable.

Of course this is all contingent on what I said before:
> Assuming logic exists to _atomically_ add an admitted pod to the active pod list before starting the next admission loop, I believe this should be sufficient.

Maybe there's good reason not to do this?

Reply:
> (1) in the active pods list; or

The bug is that the active pods list does not include force deleted pods that may take tens of minutes to complete that may still have running containers that are pinned to certain CPUs.

Reply:
Is that a problem though? I would think we can safely remove the CPUs assigned to them then, and the next time around the reconcile loop, their containers will be moved onto non-exclusive CPUs.

Reply:
Ah, nevermind, the reconcile loop also relies on the activePods List, so there is no way to know we should move these deleted pods to non-exclusive CPUs.

Reply:
On thing to consider here -- while I agree that this would be nice to have -- it's not something we supported before, so not including it now would not be a regression. It should be in it's own PR.

I also still think there's a way to handle this without adding a new function (i.e. track which pods' state was removed from `removeStaleState()` and check each of it's container statuses one last time the next time it is in the reconcile loop). If any of them are still running, it will trigger the `Update()` call at the bottom of the loop to remove it's exclusive CPUs.

Reply:
@smarterclayton Do you know if the `GetActivePods()` returned force deleted pods before the refactoring? If no, for now, we can at least fix the code to have old behavior and once we will have an option to get force deleted pods, we can refactor the code again.

Reply:
It doesn’t matter if it included these pods or not before — we’ve never had logic to actually call ‚update()‘ on the containers associated with these types of pods to remove CPUs  from them if they were still running. That is the new logic we would need to add.

Reply:
Although, I guess if activePods() did include terminated (but still running) pods before, then we never would have removed stale state about them here, making us „think“ their CPUs were free (even though containers are still running on them).

Reply:
> Although, I guess if activePods() used to include terminated (but still running) pods before, then we never would have removed stale state about them here, making us „think“ their CPUs were free (even though containers are still running on them).
AFAIU/CT this is the most pressing question now. I fully agree about the general idea to restore the old (and, we learned, not covering some important cases) behaviour and iterating later to actually cover them. The former is important and urgent, the later important as well but less urgent.

Reply:
ActivePods() never returned pods that were force deleted.  So you have always been broken on that.

Reply:
In this case, can we restore at least the old behavior under the current PR, and once we will have the infrastructure to get force deleted pods we can improve the logic under resource managers.

Reply:
I opened an issue to monitor force deleted pods problem - https://github.com/kubernetes/kubernetes/issues/104099#issuecomment-891796528

Reply:
I agree this does not block this change, and probably the fix for 104099 should make the code changes here simpler (but we don't have to do it now).